Ensure payment methods aren't detached on local or staging sites when the new checkout experience is enabled

james-allan opened the pull request:
Fixes #3358 

## Changes proposed in this Pull Request:

In https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2672, we added a new function that prevents payment methods from being detached from the Stripe customer on non-production sites. 

That new function was only added to the `detach_payment_method_from_customer()` function which is only called on sites with the legacy checkout enabled. 

When a WP user is deleted, WooCommerce calls the `wc_delete_user_data()` function which deletes all the user's payment tokens ([code ref](https://github.com/woocommerce/woocommerce/blob/9.1.0/plugins/woocommerce/includes/wc-user-functions.php#L907-L909)).

When the payment token is deleted, the WC Stripe plugin hooks in and in the `woocommerce_payment_token_deleted()` function ([code ref](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8.6.0/includes/class-wc-stripe-payment-tokens.php#L408-L423)), it detaches the payment methods in Stripe -- this is where the flow slits.

If `WC_Stripe_Feature_Flags::is_upe_checkout_enabled()` passes (new checkout experience) we call [`$stripe_customer->detach_payment_method()`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8.5.0/includes/class-wc-stripe-payment-tokens.php#L413), on stores with the legacy checkout we call [`$stripe_customer->delete_source()`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8.5.0/includes/class-wc-stripe-payment-tokens.php#L417) instead.

The `delete_source()` flow (legacy) eventually calls `WC_Stripe_API::detach_payment_method_from_customer()` which calls `should_detach_payment_method_from_customer()`. That's the function that checks the WP environment before detaching the source in Stripe.

`detach_payment_method()` (new checkout experience) does not lead to `should_detach_payment_method_from_customer()`, it just [detaches the source directly](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8.5.0/includes/class-wc-stripe-customer.php#L541).

**This PR fixes that.** 
**Deleting a source and deleting a payment method will use the same core API function which checks the environment before detaching the PM.** 

## Testing instructions

1. Make sure the new checkout experience is enabled. 
2. Add the following code snippet to your site. 

```php
define( WP_ENVIRONMENT_TYPE, 'local' );
```

3. Because you'll be testing in test mode, you'll need to comment out [this line](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8.5.0/includes/class-wc-stripe-api.php#L386).
    - payment methods are always deleted while in test mode.
3. In an incognito browser window purchase an order and create an account.
4. On the My Account > Payment methods page add a payment method. 
5. In your Stripe Dashboard check the customer and note the payment method is attached.

<p align="center">
<img width="739" alt="Screenshot 2024-08-21 at 2 18 30 PM" src="https://github.com/user-attachments/assets/c2ed1fb9-3b4b-4d9e-9fbe-76ecc10581b5">
</p>  

7. Go to the Users list table in your Admin dashboard. 
8. Delete the user you just created.
    - On `develop` the payment method for that user would be detached from stripe. 
    - On this branch the payment method should **not** be detached. 

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

james-allan committed:
Ensure pms aren't detached on local or staging sites when the new checkout experience is enabled

james-allan committed:
Use single function for detaching payment methods

james-allan committed:
Merge branch 'develop' into issue/3358-fix-detached-pms-on-staging-sites

james-allan committed:
Add changelog entries

james-allan commented:
@Mayisha I've request you're review as the original PR author, feel free to redirect to the porter next week. :)

Mayisha reviewed (commented):
@james-allan the code changes look good to me 👍 

However, I can not reproduce the original issue. And even in test mode with new/legacy checkout experience, the attached payment methods are not deleted from my Stripe customer when I delete the WP user 🤔

james-allan commented:
> And even in test mode with new/legacy checkout experience, the attached payment methods are not deleted from my Stripe customer when I delete the WP user 🤔

Yeah that's expected. We fixed it for the legacy checkout experience in #2672. However it wasn't fixed for the new checkout experience because the new checkout experience when it deletes tokens deleted them directly via an API function, not via `detach_payment_method_from_customer()` which has this environment logic.

I retested it today and followed these steps:

1. Set the `WP_ENVIRONMENT_TYPE` const:  `define( WP_ENVIRONMENT_TYPE, 'local' );`
1. In an incognito browser I added a product to my cart. 
2. On the checkout page choose to create an account. 
3. Complete the checkout.
4. Go to the **My Account > Payment methods** page. 
5. If you don't have a saved token, create one. 
6. View the customer in your Stripe dashboard and confirm the token is saved.

<p align="center">
<img width="765" alt="Screenshot 2024-08-29 at 2 45 54 PM" src="https://github.com/user-attachments/assets/55e04384-c561-4e05-9408-ff675ace7ed2">
</br>
</p> 

7. Search for the user you created in the Users list table in your WP dashboard. 
8. Delete the user you created. 
9. On `trunk` if you view the Stripe Dashboard and you'll notice that the toke was deleted. 

<p align="center">
<img width="767" alt="Screenshot 2024-08-29 at 2 53 28 PM" src="https://github.com/user-attachments/assets/6efe9942-d68e-41bf-8cb3-7dc909edf123">
</p> 

<p align="center">
<img width="768" alt="Screenshot 2024-08-29 at 2 54 31 PM" src="https://github.com/user-attachments/assets/23e03914-26f9-4544-8670-399c9e56b1fb">
</br>
<sup>At the bottom of the Customer page in Stripe, the token was detached event</sup>
</p> 

12. Checkout this branch and repeat the steps above.
   - ❗ You'll need to comment out this line https://github.com/woocommerce/woocommerce-gateway-stripe/blob/release/8.6.1/includes/class-wc-stripe-api.php#L397 because it causes the tokens to always be deleted in test mode.
13. After deleting the customer, check the Stripe dashboard and confirm the PM hasn't been detached/deleted. 

<p align="center">
<img width="768" alt="Screenshot 2024-08-29 at 3 01 43 PM" src="https://github.com/user-attachments/assets/871965d9-7fae-4921-8c61-a3883f6ca3cc">
</br>
</p>

Mayisha committed:
Merge branch 'develop' into issue/3358-fix-detached-pms-on-staging-sites

Mayisha reviewed (approved):
Thanks @james-allan for the details [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3370#issuecomment-2316720747). I am approving this PR as the changes are correct 💯 

I think there is something weird going on in my test site.

> And even in test mode with new/legacy checkout experience, the attached payment methods are not deleted from my Stripe customer when I delete the WP user 🤔

This is still happening to me in `test mode`. But in `test mode` the payment methods should always be detached (in this branch). 

> 9. On trunk if you view the Stripe Dashboard and you'll notice that the toke was deleted.

I can still see the payment method in my Stripe dashboard attached to the customer in `trunk`. What I found is the payment methods are not detached because the code returns from [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-stripe-customer.php#L566-L568) when I am testing. I am trying to figure out what's going on there. When I comment out this check everything works as described 🤔

james-allan commented:
> What I found is the payment methods are not detached because the code returns from [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-stripe-customer.php#L566-L568) when I am testing.

Oh I think I know what's happening. 😆 

I think your admin user might not have a Stripe customer ID. 

The flow is: 

1. When a user is deleted WC calls `wc_delete_user_data()` and [deletes all tokens](https://github.com/woocommerce/woocommerce/blob/9.1.0/plugins/woocommerce/includes/wc-user-functions.php#L907-L909). 
2. When a token is deleted, it triggers the `woocommerce_payment_token_deleted` action. 
3. We hook onto that and in `woocommerce_payment_token_deleted()` we create a [`WC_Stripe_Customer` instance using the **current user ID** ❌](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/a5d3ce01c1eec536a1b3f8b0468d82ffa337bc4a/includes/class-wc-stripe-payment-tokens.php#L409). 

This doesn't actually lead to an issue on my site because my admin user has a Stripe customer ID, and the eventual API request just detaches the payment method - it's not specific to my admin account or anything (ie `"payment_methods/$payment_method_id/detach"`).

james-allan committed:
Use the token's user ID rather than the current user when a token is deleted

james-allan committed:
Revert "Use the token's user ID rather than the current user when a token is deleted"

This reverts commit 524752e33cff1283c64c37c523208ee5f794cf33.

james-allan commented:
> This doesn't actually lead to an issue on my site because my admin user has a Stripe customer ID, and the eventual API request just detaches the payment method - it's not specific to my admin account or anything (ie "payment_methods/$payment_method_id/detach").

Given this PR has an approving PR review and it's a pain to test. I'll file a different issue and PR to fix this.

james-allan committed:
Merge branch 'develop' into issue/3358-fix-detached-pms-on-staging-sites
